### Import Libraries

In [2]:
import numpy as np
import os
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib 
%matplotlib inline

### Load Datasets

In [4]:
os.getcwd()

'C:\\Users\\HP'

In [5]:
os.chdir("C:\\Users\\HP\\Desktop\\New folder (2)\\python pro\\Data For Python Assessment")

##### load employee data

In [24]:

emp_df = pd.read_csv('employee_data.csv')

In [25]:
# Quick view of dataset

emp_df.head()

AGENT_ID      AGENT_NAME DATE_OF_JOINING          ADDRESS_LINE1  \
0  AGENT00001       Ray Johns      1993-06-05       1402 Maggies Way   
1  AGENT00002   Angelo Borjon      2005-12-27         414 Tanya Pass   
2  AGENT00003  Candy Spellman      2003-09-02    606 National Street   
3  AGENT00004      Mary Smith      2004-09-23  235 Hugh Thomas Drive   
4  AGENT00005    Mildred Diaz      2011-06-21  3426 Broadview Street   

  ADDRESS_LINE2              CITY STATE  POSTAL_CODE  EMP_ROUTING_NUMBER  \
0           NaN  Waterbury Center    VT         5677            34584958   
1           NaN       Panama City    FL        32404           107363763   
2          #306      Fayetteville    AR        72701            81744097   
3           NaN       Panama City    FL        32404            67563771   
4           NaN        Montgomery    AL        36110           114951317   

      EMP_ACCT_NUMBER  
0  HKUN51252328472585  
1  OPIS19290040088204  
2  YSCJ67489688482590  
3  ZANG21285355574581  
4  DZFS82244494451134

In [26]:
#Shape of dataframe

emp_df.shape

(1200, 10)

In [27]:
#information of the employee data 

emp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   AGENT_ID            1200 non-null   object
 1   AGENT_NAME          1200 non-null   object
 2   DATE_OF_JOINING     1200 non-null   object
 3   ADDRESS_LINE1       1200 non-null   object
 4   ADDRESS_LINE2       179 non-null    object
 5   CITY                1193 non-null   object
 6   STATE               1200 non-null   object
 7   POSTAL_CODE         1200 non-null   int64 
 8   EMP_ROUTING_NUMBER  1200 non-null   int64 
 9   EMP_ACCT_NUMBER     1200 non-null   object
dtypes: int64(2), object(8)
memory usage: 93.9+ KB


##### Load Insurance data

In [28]:
insurance_df = pd.read_csv("insurance_data.csv")

In [29]:
# quick view of insurance data

insurance_df.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   
2  2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969    2019-10-03   
3  2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292    2016-11-29   
4  2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204    2011-12-26   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-13  2020-05-18       Property          141.71         26000  ...   
2  2020-05-21  2020-05-26       Property          157.24         13000  ...   
3  2020-05-14  2020-05-19         Health          172.87         16000  ...   
4  2020-05-17  2020-05-22         Travel           88.53          3000  ...   

  CLAIM_STATUS INCIDENT_SEVERITY AUTHORITY_CONTACTED ANY_INJURY  \
0            A        Major Loss              Police          0   
1            A        Total Loss           Ambulance          1   
2            A        Total Loss              Police          0   
3            A        Minor Loss           Ambulance          0   
4            A        Major Loss              Police          0   

  POLICE_REPORT_AVAILABLE  INCIDENT_STATE   INCIDENT_CITY  \
0                       1              GA        Savannah   
1                       0              AL      Montgomery   
2                       1              CO  Grand Junction   
3                       0              GA        Savannah   
4                       1              TN       Nashville   

  INCIDENT_HOUR_OF_THE_DAY    AGENT_ID  VENDOR_ID  
0                        4  AGENT00413  VNDR00556  
1                        0  AGENT00769  VNDR00592  
2                       19  AGENT00883  VNDR00031  
3                       12  AGENT00278  VNDR00075  
4                       18  AGENT00636  VNDR00472  

[5 rows x 38 columns]

In [30]:
#Shape of dataframe

insurance_df.shape

(10000, 38)

In [31]:
# information of insurance dataframe
insurance_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TXN_DATE_TIME             10000 non-null  object 
 1   TRANSACTION_ID            10000 non-null  object 
 2   CUSTOMER_ID               10000 non-null  object 
 3   POLICY_NUMBER             10000 non-null  object 
 4   POLICY_EFF_DT             10000 non-null  object 
 5   LOSS_DT                   10000 non-null  object 
 6   REPORT_DT                 10000 non-null  object 
 7   INSURANCE_TYPE            10000 non-null  object 
 8   PREMIUM_AMOUNT            10000 non-null  float64
 9   CLAIM_AMOUNT              10000 non-null  int64  
 10  CUSTOMER_NAME             10000 non-null  object 
 11  ADDRESS_LINE1             10000 non-null  object 
 12  ADDRESS_LINE2             1495 non-null   object 
 13  CITY                      9946 non-null   object 
 14  STATE  

###### Load Vendor data

In [32]:
vendor_df = pd.read_csv("vendor_data.csv")

In [33]:
vendor_df.head()

VENDOR_ID              VENDOR_NAME                   ADDRESS_LINE1  \
0  VNDR00001  King, Proctor and Jones        2027 North Shannon Drive   
1  VNDR00002               Garcia Ltd          5701 East Shirley Lane   
2  VNDR00003               Cherry LLC            1217 Cottondale Road   
3  VNDR00004              Mays-Benson  227 West Montgomery Cross Road   
4  VNDR00005               Wilson PLC            23 North Hill Street   

  ADDRESS_LINE2          CITY STATE  POSTAL_CODE  
0            #5  Fayetteville    AR        72703  
1           NaN    Montgomery    AL        36117  
2           NaN    Montgomery    AL        36109  
3          #736      Savannah    GA        31406  
4           NaN     Nashville    TN        37210

In [34]:
#shape of vendor dataframe

vendor_df.shape

(600, 7)

In [35]:
# information of vendor dataframe

vendor_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   VENDOR_ID      600 non-null    object
 1   VENDOR_NAME    600 non-null    object
 2   ADDRESS_LINE1  600 non-null    object
 3   ADDRESS_LINE2  97 non-null     object
 4   CITY           594 non-null    object
 5   STATE          600 non-null    object
 6   POSTAL_CODE    600 non-null    int64 
dtypes: int64(1), object(6)
memory usage: 32.9+ KB


# TASK 1

### Merge the 3 dataset and create 1 view of data.

In [36]:
# merge insurance_data.csv and employee_data.csv on agent_id

df_1 = pd.merge(insurance_df,emp_df,on="AGENT_ID",how ="left")

In [37]:
# merge df_1 and vendor_data.csv on VENDOR_ID

df = pd.merge(df_1,vendor_df,on ="VENDOR_ID" ,how='left')

In [38]:
df.head()

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0  2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468    2015-06-23   
1  2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594    2018-04-21   
2  2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969    2019-10-03   
3  2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292    2016-11-29   
4  2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204    2011-12-26   

      LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  ...  \
0  2020-05-16  2020-05-21         Health          157.13          9000  ...   
1  2020-05-13  2020-05-18       Property          141.71         26000  ...   
2  2020-05-21  2020-05-26       Property          157.24         13000  ...   
3  2020-05-14  2020-05-19         Health          172.87         16000  ...   
4  2020-05-17  2020-05-22         Travel           88.53          3000  ...   

  STATE_y POSTAL_CODE_y EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0      TN         37217           35441155  LRZO53254369804934   
1      AZ         85301           44055451  XBYG14382904542874   
2      AL         36105           55770246  WRXH14597428306333   
3      OK         73071           87652976  YEPN16077902472499   
4      MD         21061           76119019  HZBK86544578578547   

                        VENDOR_NAME              ADDRESS_LINE1 ADDRESS_LINE2  \
0                        Ramsey Inc       3114 West 20th Court           NaN   
1  Morales, Campbell and Washington      42 West Louise Street             B   
2                       Johnson Ltd    2619 North Quality Lane          #315   
3                  Blackwell-Foster          3609 Meadow Drive           NaN   
4                       Evans-Scott  4001 19th Place Northeast           NaN   

           CITY  STATE  POSTAL_CODE  
0   Panama City     FL      32405.0  
1  Fayetteville     AR      72701.0  
2  Fayetteville     AR      72703.0  
3     Nashville     TN      37215.0  
4    Washington     DC      20018.0  

[5 rows x 53 columns]

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 0 to 9999
Data columns (total 53 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TXN_DATE_TIME             10000 non-null  object 
 1   TRANSACTION_ID            10000 non-null  object 
 2   CUSTOMER_ID               10000 non-null  object 
 3   POLICY_NUMBER             10000 non-null  object 
 4   POLICY_EFF_DT             10000 non-null  object 
 5   LOSS_DT                   10000 non-null  object 
 6   REPORT_DT                 10000 non-null  object 
 7   INSURANCE_TYPE            10000 non-null  object 
 8   PREMIUM_AMOUNT            10000 non-null  float64
 9   CLAIM_AMOUNT              10000 non-null  int64  
 10  CUSTOMER_NAME             10000 non-null  object 
 11  ADDRESS_LINE1_x           10000 non-null  object 
 12  ADDRESS_LINE2_x           1495 non-null   object 
 13  CITY_x                    9946 non-null   object 
 14  STATE_x

# TASK 2

#### find the Top 3 Insurance Type?

In [40]:
df['INSURANCE_TYPE'].value_counts()[:3]

Property    1692
Mobile      1692
Health      1690
Name: INSURANCE_TYPE, dtype: int64

###### Property,Mobile and Health are the top three types of insurance where we get the most insurance claims.

# TASK 3

###### find Top 5 States where we are getting most insurance claims for customer belonging to HIGH(H) risk segment?

In [157]:
df[df['RISK_SEGMENTATION']=='H'].STATE.value_counts()[:3]

CA    112
AR     85
CO     78
Name: STATE, dtype: int64

# TASK 4

##### create a new variable “COLOCATION” which will have following values IF Customer State == Incident State == Agent Address State THEN 1 ELSE 0 Find the mean of this new column

In [42]:
df['COLOCATION'] = np.where((df['STATE_x']== df['INCIDENT_STATE']) & (df['INCIDENT_STATE']== df['STATE_y']),1,0)

In [43]:
df['COLOCATION']

0       0
1       0
2       0
3       0
4       0
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: COLOCATION, Length: 10000, dtype: int32

# TASK 5

In [55]:
df.loc[:,['AUTHORITY_CONTACTED','POLICE_REPORT_AVAILABLE']]

AUTHORITY_CONTACTED  POLICE_REPORT_AVAILABLE
0                 Police                        1
1              Ambulance                        0
2                 Police                        1
3              Ambulance                        0
4                 Police                        1
...                  ...                      ...
9995              Police                        1
9996              Police                        0
9997              Police                        1
9998              Police                        1
9999           Ambulance                        0

[10000 rows x 2 columns]

##### Data entry error was detected in the data and you are required to correct it. If for anyclaim transaction “AUTHORITY_CONTACTED” is NOT “Police” and POLICE_AVAILABLE == 1Then Update “AUTHORITY_CONTACTED” to Police.

In [48]:
df.loc[(df['POLICE_REPORT_AVAILABLE']==1)&(df['AUTHORITY_CONTACTED']!='Police'),"AUTHORITY_CONTACTED"]="Police"

In [54]:
df

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER  \
0     2020-06-01 00:00:00    TXN00000001   A00003822   PLC00008468   
1     2020-06-01 00:00:00    TXN00000002   A00008149   PLC00009594   
2     2020-06-01 00:00:00    TXN00000003   A00003172   PLC00007969   
3     2020-06-01 00:00:00    TXN00000004   A00007572   PLC00009292   
4     2020-06-01 00:00:00    TXN00000005   A00008173   PLC00000204   
...                   ...            ...         ...           ...   
9995  2021-06-30 00:00:00    TXN00009996   A00006029   PLC00007245   
9996  2021-06-30 00:00:00    TXN00009997   A00003555   PLC00002590   
9997  2021-06-30 00:00:00    TXN00009998   A00004230   PLC00006613   
9998  2021-06-30 00:00:00    TXN00009999   A00006694   PLC00001781   
9999  2021-06-30 00:00:00    TXN00010000   A00000390   PLC00005197   

     POLICY_EFF_DT     LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  \
0       2015-06-23  2020-05-16  2020-05-21         Health          157.13   
1       2018-04-21  2020-05-13  2020-05-18       Property          141.71   
2       2019-10-03  2020-05-21  2020-05-26       Property          157.24   
3       2016-11-29  2020-05-14  2020-05-19         Health          172.87   
4       2011-12-26  2020-05-17  2020-05-22         Travel           88.53   
...            ...         ...         ...            ...             ...   
9995    2014-04-11  2021-06-25  2021-06-26          Motor          105.75   
9996    2015-08-25  2021-06-12  2021-06-17           Life           64.59   
9997    2018-10-11  2021-06-18  2021-06-23         Health          183.50   
9998    2016-01-29  2021-06-16  2021-06-21          Motor          113.46   
9999    2014-09-17  2021-06-21  2021-06-24         Mobile            8.22   

      CLAIM_AMOUNT  ... POSTAL_CODE_y EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0             9000  ...         37217           35441155  LRZO53254369804934   
1            26000  ...         85301           44055451  XBYG14382904542874   
2            13000  ...         36105           55770246  WRXH14597428306333   
3            16000  ...         73071           87652976  YEPN16077902472499   
4             3000  ...         21061           76119019  HZBK86544578578547   
...            ...  ...           ...                ...                 ...   
9995          1000  ...         72703           35413882  LFOJ76068590051807   
9996         67000  ...         31405           91206790  AHEF60771774635030   
9997          8000  ...         40272          115241097  WREH18805592939657   
9998          2000  ...         72703           13995380  AAZI21883736448048   
9999           300  ...         73003           39790996  RPES05613584687957   

                           VENDOR_NAME              ADDRESS_LINE1  \
0                           Ramsey Inc       3114 West 20th Court   
1     Morales, Campbell and Washington      42 West Louise Street   
2                          Johnson Ltd    2619 North Quality Lane   
3                     Blackwell-Foster          3609 Meadow Drive   
4                          Evans-Scott  4001 19th Place Northeast   
...                                ...                        ...   
9995                               NaN                        NaN   
9996          Newton, Hodges and Henry      32512 Christine Drive   
9997                    Reese-Mckinney         20 Garfield Street   
9998       Harmon, Manning and Bradley           1823 Dinger Road   
9999                    Foster-Sanchez        4255 Tanager Common   

      ADDRESS_LINE2          CITY STATE  POSTAL_CODE  COLOCATION  
0               NaN   Panama City    FL      32405.0           0  
1                 B  Fayetteville    AR      72701.0           0  
2              #315  Fayetteville    AR      72703.0           0  
3               NaN     Nashville    TN      37215.0           0  
4               NaN    Washington    DC      20018.0           0  
...             ...           ...   ...          ...  

# TASK 6

###### calculate claim_deviation CLAIM_DEVIATION = AVG_CLAIM_AMOUNT_FOR_LAST_30DAYS (same insurance type)/ CURRENT_CLAIM_AMOUNT

In [56]:
# current_date

df['TXN_DATE_TIME'].max()

'2021-06-30 00:00:00'

In [57]:
curr_date = pd.to_datetime(df['TXN_DATE_TIME'].max())

In [59]:
# convert 'TXN_DATE_TIME' to datetime

df['TXN_DATE_TIME'] = pd.to_datetime(df['TXN_DATE_TIME'])

In [60]:
# Average claim amount for last 30 days
# last 30 days data

L_data  = df.loc[(df['TXN_DATE_TIME'] > '2021-05-30') & (df['TXN_DATE_TIME'] < '2021-06-30')]

In [66]:
# avg claim 

avg_claim = L_data.groupby('INSURANCE_TYPE').mean()['CLAIM_AMOUNT'].to_dict()

In [67]:
avg_claim

{'Health': 10904.411764705883,
 'Life': 52242.857142857145,
 'Mobile': 378.35820895522386,
 'Motor': 5573.770491803279,
 'Property': 24627.906976744187,
 'Travel': 2800.0}

###### If CLAIM_DEVIATION < 0.5 THEN CLAIM_DEVIATION = 1 ELSE 0 If there is less than 30 days of transaction history THEN -1

In [71]:
def claim_deviation(row):
    if (curr_date - row['TXN_DATE_TIME'])<pd.Timedelta(days=30):
        return -1
    ClAIM_DEVIATION = avg_claim[row['INSURANCE_TYPE']] / row['CLAIM_AMOUNT']
    if ClAIM_DEVIATION < 0.5:
        return 1
    else:
        return 0

In [75]:
df['CLAIM_DEVIATION'] = df.apply(claim_deviation,axis=1)
df

TXN_DATE_TIME TRANSACTION_ID CUSTOMER_ID POLICY_NUMBER POLICY_EFF_DT  \
0       2020-06-01    TXN00000001   A00003822   PLC00008468    2015-06-23   
1       2020-06-01    TXN00000002   A00008149   PLC00009594    2018-04-21   
2       2020-06-01    TXN00000003   A00003172   PLC00007969    2019-10-03   
3       2020-06-01    TXN00000004   A00007572   PLC00009292    2016-11-29   
4       2020-06-01    TXN00000005   A00008173   PLC00000204    2011-12-26   
...            ...            ...         ...           ...           ...   
9995    2021-06-30    TXN00009996   A00006029   PLC00007245    2014-04-11   
9996    2021-06-30    TXN00009997   A00003555   PLC00002590    2015-08-25   
9997    2021-06-30    TXN00009998   A00004230   PLC00006613    2018-10-11   
9998    2021-06-30    TXN00009999   A00006694   PLC00001781    2016-01-29   
9999    2021-06-30    TXN00010000   A00000390   PLC00005197    2014-09-17   

         LOSS_DT   REPORT_DT INSURANCE_TYPE  PREMIUM_AMOUNT  CLAIM_AMOUNT  \
0     2020-05-16  2020-05-21         Health          157.13          9000   
1     2020-05-13  2020-05-18       Property          141.71         26000   
2     2020-05-21  2020-05-26       Property          157.24         13000   
3     2020-05-14  2020-05-19         Health          172.87         16000   
4     2020-05-17  2020-05-22         Travel           88.53          3000   
...          ...         ...            ...             ...           ...   
9995  2021-06-25  2021-06-26          Motor          105.75          1000   
9996  2021-06-12  2021-06-17           Life           64.59         67000   
9997  2021-06-18  2021-06-23         Health          183.50          8000   
9998  2021-06-16  2021-06-21          Motor          113.46          2000   
9999  2021-06-21  2021-06-24         Mobile            8.22           300   

      ... EMP_ROUTING_NUMBER     EMP_ACCT_NUMBER  \
0     ...           35441155  LRZO53254369804934   
1     ...           44055451  XBYG14382904542874   
2     ...           55770246  WRXH14597428306333   
3     ...           87652976  YEPN16077902472499   
4     ...           76119019  HZBK86544578578547   
...   ...                ...                 ...   
9995  ...           35413882  LFOJ76068590051807   
9996  ...           91206790  AHEF60771774635030   
9997  ...          115241097  WREH18805592939657   
9998  ...           13995380  AAZI21883736448048   
9999  ...           39790996  RPES05613584687957   

                           VENDOR_NAME              ADDRESS_LINE1  \
0                           Ramsey Inc       3114 West 20th Court   
1     Morales, Campbell and Washington      42 West Louise Street   
2                          Johnson Ltd    2619 North Quality Lane   
3                     Blackwell-Foster          3609 Meadow Drive   
4                          Evans-Scott  4001 19th Place Northeast   
...                                ...                        ...   
9995                               NaN                        NaN   
9996          Newton, Hodges and Henry      32512 Christine Drive   
9997                    Reese-Mckinney         20 Garfield Street   
9998       Harmon, Manning and Bradley           1823 Dinger Road   
9999                    Foster-Sanchez        4255 Tanager Common   

     ADDRESS_LINE2          CITY STATE POSTAL_CODE  COLOCATION  \
0              NaN   Panama City    FL     32405.0           0   
1                B  Fayetteville    AR     72701.0           0   
2             #315  Fayetteville    AR     72703.0           0   
3              NaN     Nashville    TN     37215.0           0   
4              NaN    Washington    DC     20018.0           0   
...            ...           ...   ...         ...         ...   
9995           NaN           NaN   NaN         NaN           0   
9996           NaN    Union City    CA     94587.0           0   
9997           NaN       Needham    MA      2492.0           0   
9998           NaN     Annapolis    MD   

In [77]:
df['CLAIM_DEVIATION'].value_counts()

 0    9216
-1     784
Name: CLAIM_DEVIATION, dtype: int64

* All the claim deviation values are graiter than 0.5
* 784 claims have less than 30 days transaction historu


# TASK 7

###### Find All Agents who have worked on more than 2 types of Insurance Claims. Sort them by Total Claim Amount Approved under them in descending order

In [93]:
# group by agent id

agents =df.groupby('AGENT_ID')['INSURANCE_TYPE'].nunique()
agents


AGENT_ID
AGENT00001    4
AGENT00002    2
AGENT00003    6
AGENT00004    6
AGENT00005    4
             ..
AGENT01196    4
AGENT01197    4
AGENT01198    4
AGENT01199    5
AGENT01200    6
Name: INSURANCE_TYPE, Length: 1200, dtype: int64

In [87]:
top_agents = agents >2 
top_agents

AGENT_ID
AGENT00001     True
AGENT00002    False
AGENT00003     True
AGENT00004     True
AGENT00005     True
              ...  
AGENT01196     True
AGENT01197     True
AGENT01198     True
AGENT01199     True
AGENT01200     True
Name: INSURANCE_TYPE, Length: 1200, dtype: bool

In [94]:
top_agentd = top_agents[top_agents].index

In [111]:
TOP_AGENTS = df.groupby(['AGENT_ID','AGENT_NAME']).sum().loc[top_agentd][['CLAIM_AMOUNT']].sort_values('CLAIM_AMOUNT',ascending=False)

In [112]:
TOP_AGENTS

CLAIM_AMOUNT
AGENT_ID   AGENT_NAME                    
AGENT00807 Don Filkins             528800
AGENT00679 Clara Barnett           489000
AGENT00771 Roger Burns             422100
AGENT00125 Salvador Soriano        400400
AGENT00789 Alison Baron            392900
...                                   ...
AGENT00910 Vincent Ragan            15500
AGENT00689 Connie Hunt              13700
AGENT00732 Sylvia Tran              11300
AGENT00706 Laura Staggs             10800
AGENT00604 Thelma Salinas            5500

[1164 rows x 1 columns]

* There are 1164 Agents who have worked on more than 2 types of Insurance Claims.
* Don Filkins is the Top most agent with highest claim amount

# TASK 8

###### Mobile & Travel Insurance premium are discounted by 10% Health and Property Insurance premium are increased by 7% Life and Motor Insurance premium are marginally increased by 2% What will be overall change in % of the Premium Amount Collected for all these Customer?

In [140]:
# total premiun amount by insurance types

premium = df.groupby('INSURANCE_TYPE')['PREMIUM_AMOUNT'].sum().to_dict()
premium

{'Health': 252455.73,
 'Life': 125621.22,
 'Mobile': 15191.42,
 'Motor': 165391.98,
 'Property': 202285.83,
 'Travel': 124139.77}

In [141]:
#Total premium amount early

sum_premium = sum(premium.values())
sum_premium

885085.95

In [143]:
# new premium amount

pre_new = premium.copy()

In [144]:
# Mobile & Travel Insurance premium are discounted by 10% 

pre_new['Mobile']*=0.9
pre_new['Travel']*=0.9

In [145]:
# Health and Property Insurance premium are increased by 7%

pre_new['Health']*=1.07
pre_new['Property']*=1.07

In [146]:
# Life and Motor Insurance premium are marginally increased by 2%

pre_new['Motor']*=1.02
pre_new['Life']*=1.02

In [148]:
# New Premium amount

pre_new

{'Health': 270127.6311,
 'Life': 128133.6444,
 'Mobile': 13672.278,
 'Motor': 168699.81960000002,
 'Property': 216445.8381,
 'Travel': 111725.793}

In [150]:
# sum of new premium amount

new_sum = sum(pre_new.values())
new_sum

908805.0042000001

In [152]:
# Total difference in new premium and early premmium amount

dif =new_sum-sum_premium
dif

23719.05420000013

In [154]:
# overall change in % of the Premium Amount

overall_change = (dif/sum_premium)*100
overall_change

2.6798588543858517

* Overall change in Premium amount collected is 2.68%

# TASK 9

 ##### Find the percentage of customers are eligible for discount 

In [159]:
# Eligibility condition: IF CUSTOMER_TENURE > 60 AND EMPLOYMENT_STATUS = “N”AND NO_OF_FAMILY_MEMBERS >=4 THEN 1 ELSE 0

df['ELIGIBLE_FOR_DISCOUNT'] = np.where((df['TENURE'] > 60) & (df['EMPLOYMENT_STATUS'] =="N") & (df['NO_OF_FAMILY_MEMBERS']>=4),1,0)
df['ELIGIBLE_FOR_DISCOUNT']

0       0
1       0
2       0
3       0
4       0
       ..
9995    0
9996    0
9997    0
9998    0
9999    0
Name: ELIGIBLE_FOR_DISCOUNT, Length: 10000, dtype: int32

In [160]:
df['ELIGIBLE_FOR_DISCOUNT'].value_counts()

0    9701
1     299
Name: ELIGIBLE_FOR_DISCOUNT, dtype: int64

In [161]:
df['ELIGIBLE_FOR_DISCOUNT'].mean()

0.0299

* 2.99 % of customers are eligible for the discount 

# TASK 10
###### check Claim Velocity

In [162]:
#current date

curr_date

Timestamp('2021-06-30 00:00:00')

In [172]:
# claims in last 30 days

L_data  = df.loc[(df['TXN_DATE_TIME'] > '2021-05-30') & (df['TXN_DATE_TIME'] < '2021-06-30')]
L_30 = pd.DataFrame.from_dict(L_data[L_data['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['CLAIMS_IN_LAST30DAYS'])

In [173]:
# claims in last 3 days

L_3days = df.loc[(df['TXN_DATE_TIME'] > curr_date - pd.Timedelta(days=4)) & (df['TXN_DATE_TIME'] < '2021-06-30')]
L_3 = pd.DataFrame.from_dict(L_3days[L_3days['CLAIM_STATUS']=='A'].groupby('INSURANCE_TYPE').size().to_dict(),orient='index',columns=['CLAIMS_IN_LAST3DAYS'])

In [177]:
# All claims 

total_claims = L_30.join(L_3)

In [179]:
total_claims['CLAIM_VELOCITY'] = total_claims['CLAIMS_IN_LAST30DAYS']/total_claims['CLAIMS_IN_LAST3DAYS']

In [180]:
total_claims

CLAIMS_IN_LAST30DAYS  CLAIMS_IN_LAST3DAYS  CLAIM_VELOCITY
Health                     128                   14        9.142857
Life                       128                   12       10.666667
Mobile                     127                    9       14.111111
Motor                      117                   12        9.750000
Property                   125                   12       10.416667
Travel                     123                   14        8.785714

# TASK 11
###### Find all low performing agents i.e. employees who are in the bottom 5 percentile based on Claims worked by them.

In [183]:
# Lowest performing agents

Low = df.groupby('AGENT_ID').sum().quantile(0.05)['CLAIM_AMOUNT']
Low

30395.0

In [186]:
# Agent Data

agent = df.groupby('AGENT_ID').sum()

In [189]:
#Low performing agent data

Low_Performing_Agents = agent[agent['CLAIM_AMOUNT'] <= Low][['CLAIM_AMOUNT']]

Low_Performing_Agents

CLAIM_AMOUNT
AGENT_ID                
AGENT00006         22400
AGENT00014          7400
AGENT00016         27000
AGENT00085         25700
AGENT00088         27700
AGENT00091         30000
AGENT00098         28200
AGENT00134         22000
AGENT00149         28900
AGENT00156         13200
AGENT00172         22200
AGENT00185         16100
AGENT00186         27700
AGENT00188         28400
AGENT00197         18500
AGENT00252         20000
AGENT00263         11500
AGENT00275         16200
AGENT00279         28000
AGENT00300         12600
AGENT00310         28800
AGENT00316         22200
AGENT00351         21700
AGENT00425         17800
AGENT00442         30300
AGENT00536         20000
AGENT00562         21100
AGENT00571         24200
AGENT00585         16600
AGENT00590         22200
AGENT00604          5500
AGENT00621          5300
AGENT00628          6100
AGENT00672         21300
AGENT00689         13700
AGENT00706         10800
AGENT00721          9000
AGENT00732         11300
AGENT00773         15000
AGENT00780         11300
AGENT00790          6000
AGENT00804         20000
AGENT00805         22600
AGENT00830         23500
AGENT00841         17800
AGENT00856         24200
AGENT00894         27600
AGENT00910         15500
AGENT00969         28300
AGENT00985         30000
AGENT01021         24300
AGENT01037          5600
AGENT01052         28000
AGENT01077          4200
AGENT01107         24700
AGENT01128         28000
AGENT01131         27500
AGENT01151         28000
AGENT01154           800
AGENT01179         27300

In [191]:
Low_Performing_Agents.count()

CLAIM_AMOUNT    60
dtype: int64

* There are 60 agents are in the bottom 5 percentile

# TASK 12
###### Business wants to find all Suspicious Employees (Agents).

In [207]:
Condition = (df['CLAIM_STATUS']=='A')&(df['RISK_SEGMENTATION']=='H')&(df['INCIDENT_SEVERITY']=='Major Loss')

In [208]:
critieria = df[Condition]

In [195]:
= np.where((df['CLAIM_STATUS']=='A')&(df['RISK_SEGMENTATION']=='H')&(df['INCIDENT_SEVERITY']=='Major Loss'),1,0)

In [220]:
suspi_df = critieria.groupby("AGENT_ID").sum()


In [223]:
sus_agents = suspi_df[suspi_df['CLAIM_AMOUNT']>=15000]
sus_agents['AGENTS'] = sus_agents.index

C:\Users\HP\AppData\Local\Temp/ipykernel_11060/1601519985.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sus_agents['AGENTS'] = sus_agents.index


In [226]:
# suspicious agents count

suspicious_agents=sus_agents['AGENTS'].count()
suspicious_agents

149

In [227]:
#total agents count

emp_count = emp_df['AGENT_ID'].count()

In [229]:
# percentage of suspicious agents count
perc = suspicious_agents/emp_count
perc

0.12416666666666666

* 12.41 percentage of agents are suspicious